File: Depth-First Search/Graph/LET_no797_All Paths From Source to Target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> res;
        vector<int> temp = {0};
        dfs(res, temp, 0, graph);
        return res;
    }
    
    void dfs(vector<vector<int>>& res, vector<int>& temp, int start, vector<vector<int>>& graph) {
        if (start == graph.size() - 1) {
            res.push_back(temp);
            return;
        }
        
        for (int i : graph[start]) {
            temp.push_back(i);
            dfs(res, temp, i, graph);
            temp.pop_back();
        }
    }
};

File: Depth-First Search/Graph/LET_no797_All Paths From Source to Target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        res = []
        self.dfs(res, [0], 0, graph)
        return res
    
    def dfs(self, res, temp, start, graph):
        if start == len(graph) - 1:
            res.append(temp[:])
            return
        
        for i in graph[start]:
            temp.append(i)
            self.dfs(res, temp, i, graph)
            temp.pop()